Login rework: Username/Password screen for selfhosted

hypest opened the pull request:
This PR introduces the username+password screen for the selfhosted flow. It supports both selfhosted and WPCOM sites.

The final visual touches, the messages and final texts will be done in a future PR.

To test:
* With the app signed out or its data cleaned, launch the app
* Tap on the "Login" button to go to the email screen
* Tap on the "Log in to your site by entering your site address instead." link on the lower left to go to the site address input screen
* Enter the URL of your selfhosted or wpcom site and tap on "NEXT"
* Enter your username and password in the respective input fields and tap on "NEXT"
* Notice the app do some work (progress popup) and jumping to the epilogue screen on success

nbradbury commented:
I'm guessing this is already a future task, but if you log in to your .org site, switch to the Me tab, then select "Log in to WordPress.com," you end up going through the same login flow that you started with. Which is confusing because it does a lot more than enable logging into wp.com.

nbradbury commented:
I don't think we want to include `xmlrpc.php` here.

![screenshot_1497908243](https://user-images.githubusercontent.com/3903757/27307357-56e49476-5517-11e7-9067-8a209f312f1a.png)

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginSiteAddressFragment.java:
```diff
@@ -198,7 +208,42 @@ protected void discover() {
+                new Response.Listener<SiteWPComRestResponse>() {
+                    @Override
+                    public void onResponse(SiteWPComRestResponse response) {
+                        if (response.jetpack) {
```
I have a Jetpack site at http://www.nbradbury.com/wp/ but it's not getting handled here, instead it gets caught by `onErrorResponse` below (`apiError = `unknown_blog`). Is this expected?

hypest replied:
That's actually the expected behaviour. Since `nbradbury.com` is not a (mapped) WordPress.com site, the request will fail and be handled in `onErrorResponse` as you've already observed.

If you notice the comment on [the following line](https://github.com/wordpress-mobile/WordPress-Android/pull/6089/files#diff-bbc95f0597a9d70e9bb0293ddc4c0122R219), the Jetpack condition is there to force the WordPress.com site to be treated as selfhosted instead of WPCOM. That way, user's selfhosted credentials will be accepted and the selfhosted network calls (XMLRPC) will be used. Let me know if that makes this more clear.

nbradbury commented:
Here's an odd one: after I logged in and got to the screen with the "Connect another site" and "Continue" buttons, I hit the back button. This took me to the My Site tab, which I didn't expect.

Then I rotated the device and I was back on the "Connect another site" screen. I hit the back button again and I was taken to another instance of the "Connect another site" screen. I hit back one more time and was shown My Site again.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginUsernamePasswordFragment.java:
```diff
@@ -0,0 +1,457 @@
+    }
+
+    @Override
+    public void onAttach(Context context) {
```
I think we to use want `onAttach(Activity activity)` here because `onAttach(Context context)` wasn't added until API 23.

hypest replied:
`LoginUsernamePasswordFragment` extends the support lib's Fragment class so, `onAttach(Context context)` is the variant called, even for older API versions. The `onAttach(Activity activity)` variant is already deprecated for the version we're using. I think there is no need to override the Activity variant, right?

nbradbury replied:
> LoginUsernamePasswordFragment extends the support lib's Fragment class so, onAttach(Context context) is the variant called, even for older API versions

TIL!

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginUsernamePasswordFragment.java:
```diff
@@ -0,0 +1,457 @@
+    @Override
+    public boolean onOptionsItemSelected(MenuItem item) {
+        if (item.getItemId() == R.id.help) {
+            mLoginListener.help();
```
Does `mLoginListener` need a null check?

hypest replied:
Ah, for sure! Fixed with 566f229.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginUsernamePasswordFragment.java:
```diff
@@ -0,0 +1,457 @@
+            case AUTHORIZATION_REQUIRED:
+            case NEEDS_2FA:
+                if (mIsWpcom) {
+                    mLoginListener.needs2fa(mRequestedUsername, mRequestedPassword);
```
Does `mLoginListener` need a null check?

hypest replied:
👍  Fixed with 566f229.

hypest commented:
> I'm guessing this is already a future task, but if you log in to your .org site, switch to the Me tab, then select "Log in to WordPress.com," you end up going through the same login flow that you started with. Which is confusing because it does a lot more than enable logging into wp.com.

Indeed, it's a future task to handle entering the login flows from other places in the app (there are at least 3 other starting points like that) and will be handled in future PRs. Thanks for pointing it out anyway!

hypest commented:
> I don't think we want to include xmlrpc.php here.

Right, we probably don't, it's weird and prolly confusing for the user. Thing is, there is no good "clean" alternative and that's why I left it as is.

Here are some alternatives I had considered and didn't fully like:
* Use the initial address the user inputted. That also can become confusing since the final (actual) URL can be a very different one after redirections/rewrites.
* Get the canonical URL from the site. Unfortunately, there is no XMLRPC call to get that info while logged out. We could scrape the HTML to get some URL but that's messy.
*  Manipulating the URL string itself (say, strip the `xmlrpc.php` part). Unfortunately, not bullet-proof either due to potential redirects/rewrites.

Perhaps there are more alternatives. Open to ideas on this!

BTW, this starts to feel a lot like the problem browsers also have about what to display in their address bar, especially when logging in a site.

hypest commented:
> Here's an odd one: after I logged in and got to the screen with the "Connect another site" and "Continue" buttons, I hit the back button. This took me to the My Site tab, which I didn't expect.

That's expected behaviour. The login process has actually finished by the time you reach the epilogue screen and the epilogue is purely informational, mainly to reassure the user that they logged in the intended account. There is no "back" button _on screen_. The "Continue" button and Android's Back button dismiss the screen, dropping you back into the main app.

> Then I rotated the device and I was back on the "Connect another site" screen. I hit the back button again and I was taken to another instance of the "Connect another site" screen. I hit back one more time and was shown My Site again.

That sounds like a bug. If that's OK with you, I will open a ticket for that and handle it separately as it's actually not related to the current PR.

hypest commented:
Hey @nbradbury , thanks for picking this up for review! I think I've addressed all your comments. Have a look and let me know for any additional remarks if any. Thanks!

nbradbury commented:
> Manipulating the URL string itself (say, strip the xmlrpc.php part). Unfortunately, not bullet-proof either due to potential redirects/rewrites.

I think that's the best option. I don't think redirects/rewrites are a concern, since I'm only talking about changing the URL that's **displayed**, not the one we actually use. As it is, it feels like we're exposing our plumbing to our users :)

aforcier commented:
@hypest @nbradbury FYI I'm working on adding a new action to FluxC that will do [this work](https://github.com/wordpress-mobile/WordPress-Android/pull/6089/files#diff-bbc95f0597a9d70e9bb0293ddc4c0122R214) without exposing FluxC networking logic in WPAndroid. I'm not sure if you want to wait on that to merge this PR, but certainly we should change that before these changes reach `develop`.

hypest commented:
> > Manipulating the URL string itself (say, strip the xmlrpc.php part). Unfortunately, not bullet-proof either due to potential redirects/rewrites.
>
> I think that's the best option. I don't think redirects/rewrites are a concern, since I'm only talking about changing the URL that's **displayed**, not the one we actually use. As it is, it feels like we're exposing our plumbing to our users :)

OK @nbradbury , I'll work on that and let you know, thanks!

> @hypest @nbradbury FYI I'm working on adding a new action to FluxC (as per @hypest's request) that will do this work without exposing FluxC networking logic in WPAndroid. I'm not sure if you want to wait on that to merge this PR, but certainly we should change that before these changes reach develop.

Hey @aforcier , thx for the headsup. Yeap, am aware of that effort on the FluxC side. This PR is against the feature branch anyway and I want to switch to the FluxC solution when that lands. I think the feature branch will outlive this. Since switching to the FluxC solution is actually an optimisation at this point, i'd propose we don't block the current PR on it. I will make the switch in a future PR. WDTY @nbradbury ?

hypest commented:
After some discussion with the rest of the Login rework team, I employed some sanitisation of the _input_ address and displaying that to the user. The sanitisation includes removing the scheme and `xmlrpc.php` if happened to be present. Ready for another pass @nbradbury , thanks!

nbradbury commented:
This is **much** better :)

![screenshot_1498051215](https://user-images.githubusercontent.com/3903757/27386171-fee820f6-5662-11e7-886e-2f25bd1c18cc.png)

nbradbury commented:
:shipit:
